plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/GameStateMachine.hs

```diff
@@ -0,0 +1,219 @@
+{-# LANGUAGE TypeApplications  #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE DataKinds         #-}
+{-# OPTIONS_GHC -O0 #-}
+-- | A guessing game that
+-- 
+--   * Uses a state machine to keep track of the current secret word
+--   * Uses a token to keep track of who is allowed to make a guess
+--
+module Language.PlutusTx.Coordination.Contracts.GameStateMachine(
+      startGame
+    , guess
+    , lock
+    , gameTokenVal
+    , gameValidator
+    ) where
+
+import qualified Data.Map                     as Map
+import           Data.Maybe                   (maybeToList)
+import qualified Data.Set                     as Set
+import qualified Data.Text                    as Text
+import qualified Language.PlutusTx            as PlutusTx
+import qualified Language.PlutusTx.Prelude    as P
+import           Ledger                       hiding (to)
+import qualified Ledger.Ada                   as Ada
+import           Ledger.Value                 (TokenName)
+import qualified Ledger.Value.TH              as V
+import qualified Ledger.Validation            as Validation
+import           Wallet
+import qualified Wallet                       as WAPI
+
+import qualified Data.ByteString.Lazy.Char8   as C
+
+import Language.PlutusTx.StateMachine
+
+data HashedString = HashedString (P.SizedByteString 32)
+
+PlutusTx.makeLift ''HashedString
+
+data ClearString = ClearString (P.SizedByteString 32)
+
+PlutusTx.makeLift ''ClearString
+
+-- | State of the guessing game
+data GameState = 
+    Initialised HashedString
+    -- ^ Initial state. In this state only the 'ForgeTokens' action is allowed.
+    | Locked TokenName HashedString
+    -- ^ Funds have been locked. In this state only the 'Guess' action is 
+    --   allowed.
+
+PlutusTx.makeLift ''GameState
+
+-- | Inputs (actions)
+data GameInput =
+      ForgeToken TokenName
+    -- ^ Forge the "guess" token
+    | Guess ClearString HashedString
+    -- ^ Make a guess and lock the remaining funds using a new secret word.
+
+PlutusTx.makeLift ''GameInput
+
+gameValidator :: ValidatorScript
+gameValidator = ValidatorScript ($$(Ledger.compileScript [||
+    \(ds :: (GameState, Maybe GameInput)) (vs :: (GameState, Maybe GameInput)) (p :: PendingTx) ->

```

Comment:
Maybe we should also define a type alias for "state machine validators" in `StateMachine.hs`?